addons/mail/models/mail_guest.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import pytz
+import uuid
+from odoo.tools import consteq
+from odoo import fields, models, _
+from odoo.addons.base.models.res_partner import _tz_get
+
+class MailGuest(models.Model):
+    _name = 'mail.guest'
+    _description = "Guest"
+
+    access_token = fields.Char(
+        copy=False,
+        default=lambda self: str(uuid.uuid4()),
+        groups='base.group_system',
+        readonly=True,
+        required=True,
+    )
+    channel_ids = fields.Many2many(comodel_name='mail.channel')
+    country_id = fields.Many2one(string="Country", comodel_name='res.country')
+    lang_id = fields.Many2one(string="Language", comodel_name='res.lang')
+    name = fields.Char(required=True)
+    timezone = fields.Selection(selection=_tz_get)
+
+    def format_guests(self):
+        formatted_guests = [{
+            'id': guest.id,
+            'name': guest.name,
+        } for guest in self]
+
+        return formatted_guests
+
+    def _get_guest_from_request(self, request):
+        # Retrieve required data from cookies.
+        guest_id = request.httprequest.cookies.get('mail.guest_id')
+        guest_access_token = request.httprequest.cookies.get('mail.guest_access_token')
+        if not guest_id or not guest_access_token:
+            return self.env['mail.guest']
+
+        # Ensure the guest exists and their access token is valid.
+        guest = self.env['mail.guest'].sudo().search([('id', '=', guest_id)], limit=1)
+        if not guest or not consteq(guest.access_token, guest_access_token):
+            return self.env['mail.guest']
+
+        if not guest.timezone:
+            timezone = self._get_timezone_from_request(request)
+            if timezone:
+                guest._update_timezone(timezone)
+
+        return guest.sudo(False)
+
+    def _get_or_create_guest_from_request(self, request, response):
+        guest = self.env['mail.guest']._get_guest_from_request(request)
+        if guest:
+            return guest
+        country_code = request.session.get('geoip', {}).get('country_code', False)
+        country = self.env['res.country'].sudo().search([
+            ('code', '=', country_code)
+        ], limit=1) if country_code else self.env['res.country']
+        guest = self.sudo().create({
+            'country_id': country.id,
+            # 'lang_id': request.lang.id,
+            'lang_id': self.env['res.lang'].search([('code', '=', 'en_US')], limit=1).id,
```

Comment:
`en_US` is not necessarily enabled on a db, you have `tools.misc.get_lang` that finds a valid lang for you